step-maven-plugin/step-commons-maven-plugin/src/main/java/step/plugins/maven/AbstractUploadKeywordsPackageMojo.java

```diff
@@ -0,0 +1,125 @@
+package step.plugins.maven;
+
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugin.MojoFailureException;
+import org.apache.maven.plugins.annotations.Parameter;
+import org.apache.maven.project.MavenProject;
+import step.client.accessors.RemoteAccessors;
+import step.client.collections.remote.RemoteCollectionFactory;
+import step.core.accessors.AbstractAccessor;
+import step.functions.packages.FunctionPackage;
+import step.functions.packages.client.RemoteFunctionPackageClientImpl;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+
+public abstract class AbstractUploadKeywordsPackageMojo extends AbstractStepPluginMojo {
+
+	@Parameter(defaultValue = "${project}", readonly = true, required = true)
+	private MavenProject project;
+
+	@Parameter(property = "step-upload-keywords.custom-package-attrs", required = false)
+	private Map<String, String> customPackageAttributes;
+
+	@Override
+	public void execute() throws MojoExecutionException, MojoFailureException {
+		getLog().info("Uploading keywords package to step...");
+
+		try (RemoteFunctionPackageClientImpl remoteFunctionPackageClient = createRemoteFunctionPackageClient()) {
+			File packagedTarget = getFileToUpload();
+
+			FunctionPackage previousPackage = null;
+			Map<String, String> packageAttributes;
+
+			// init with default if the key is not explicitly defined via maven parameter
+			if (getCustomPackageAttributes() == null || getCustomPackageAttributes().isEmpty()) {
+				packageAttributes = new HashMap<>();
+				packageAttributes.put("artifactId", getProject().getArtifactId());
+			} else {
+				packageAttributes = getCustomPackageAttributes();
```

Comment:
@david-stephan @iegorov777 We might want to use custom package attributes without using them as tracking attributes. Shouldn't we always use the concatenation of groupId + artifactId as tracking attribute?

Reply:
@jeromecomte probably we can get the situation when want to use versionId additionally to groupId+artifactId
But anyway we can split the configuration parameters for "tracking attributes" and "other package attribute" and apply them independently (if it makes sense).

Reply:
@iegorov777 We aligned with @david-stephan. Here the solution we propose:

Per default if getCustomPackageAttributes() is not set, no package attribute has to be set (remove artifactId)
If getCustomPackageAttributes() is set, use it as package attribute (argument the method of newKeywordPackage())

As tracking attribute use a a custom field of the function package: FunctionPackage.customFields.tracking (step.core.accessors.AbstractIdentifiableObject#customFields) and set it to groupId + "." + artifactId. To do so, the method step.functions.packages.client.RemoteFunctionPackageClientImpl#addOrUpdateKeywordPackage might have to be extended

Only this tracking attribute should be used to search for previously uploaded packages